src/decoder_pipeline.jl

```diff
@@ -0,0 +1,246 @@
+abstract type AbstractSyndromeDecoder end
+
+"""Calls evaluate_classical_decoder for the X and Z checks independently"""
+function evaluate_decoder(d::AbstractSyndromeDecoder, nsamples, init_error, gate_error, syndrome_circuit_func, encoding_circuit_func)
+    pre_X = [sHadamard(i) for i in n-k+1:n]
+    X_error = evaluate_classical_decoder(d, nsamples, init_error, gate_error, syndrome_circuit_func, encoding_circuit_func, logicalxview, 1, d.k, pre_X)
+    Z_error = evaluate_classical_decoder(d, nsamples, init_error, gate_error, syndrome_circuit_func, encoding_circuit_func, logicalzview, d.k + 1, 2 * d.k)
+    return (X_error, Z_error)
+end
+
+"""Uses a Pauli frame simulation to evaluate each part of the CSS code (X or Z tableau)"""
+function evaluate_classical_decoder(d::AbstractSyndromeDecoder, nsamples, init_error, gate_error, syndrome_circuit_func, encoding_circuit_func, logical_view_function, guess_start, guess_stop, pre_circuit = nothing)
+    H = d.H
+    O = d.faults_matrix
+    syndrome_circuit = syndrome_circuit_func(H)
+
+    n = d.n
+    s = d.s
+    k = d.k
+
+    errors = [PauliError(i, init_error) for i in 1:n];
+
+    md = MixedDestabilizer(H)
+    
+    full_circuit = []
+
+    logview = logical_view_function(md)
+    logcirc, _ = syndrome_circuit_func(logview)
+
+    noisy_syndrome_circuit = add_two_qubit_gate_noise(syndrome_circuit, gate_error);
+    
+    for gate in logcirc
+        type = typeof(gate)
+        if type == sMRZ
+            push!(syndrome_circuit, sMRZ(gate.qubit+s, gate.bit+s))
+        else
+            push!(syndrome_circuit, type(gate.q1, gate.q2+s))
+        end
+    end
+
+    ecirc = encoding_circuit_func(syndrome_circuit)
+    if isnothing(pre_circuit)
+        full_circuit = vcat(pre_circuits, ecirc, errors, noisy_syndrome_circuit)
+    else
+        full_circuit = vcat(ecirc, errors, noisy_syndrome_circuit)
+    end
+
+    frames = PauliFrame(nframes, n+s+k, s+k)
+    pftrajectories(frames, full_circuit)
+    syndromes = pfmeasurements(frames)[:, 1:s]
+    logical_syndromes = pfmeasurements(frames)[:, s+1: s+k]
+
+    for i in 1:nsamples
+        guess = decode(d, syndromes[i])
+
+        result = (O * (guess))[guess_start:guess_stop]
+        
+        if result == logical_syndromes[i]
+            decoded += 1
+        end
+    end
+
+    return (nsamples - decoded) / nsamples
+end
+
+
+"""Type for evaluating the table decoder"""
+struct TableDecoder <: AbstractSyndromeDecoder
+    H
+    faults_matrix
+    n
+    s
```

Comment:
Let's add some documentation inline for these structs. Something like this

```suggestion
    "the stabilizer tableau defining the code"
    H
    "the faults matrix that converts a physical error vector into corresponding logical error vector"
    faults_matrix
    n
    s
```